src/pybind/mgr/cephadm/agent.py

```diff
@@ -0,0 +1,218 @@
+import cherrypy
+import json
+import socket
+import tempfile
+import threading
+import time
+
+from mgr_util import create_self_signed_cert, verify_tls_files
+from ceph.utils import datetime_now
+
+from typing import Set, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from cephadm.module import CephadmOrchestrator
+
+
+class CherryPyThread(threading.Thread):
+    def __init__(self, mgr: "CephadmOrchestrator") -> None:
+        self.mgr = mgr
+        self.cherrypy_shutdown_event = threading.Event()
+        super(CherryPyThread, self).__init__(target=self.run)
+
+    def run(self) -> None:
+        class Root(object):
```

Comment:
why not move this Root out of the run method and use it to handle the logic instead of starting new threads from a thread that cherrypy has already dispatched? Nitpick - you don't need (object) on the class definition in py3.

something like https://gist.github.com/pcuzner/429135463a88c118ac43b971c84daceb

You could then potentially expose a setting to tweak the cherrypy thread_pool to keep things simple. The default of 10 depends on your service time. In the exporter code a scrape took -18-25ms - so that would be a lot of scrapes in 10s. Whether the default is enough for this depends on your timings - I suspect they'll be fine, but worth checking you push latency to arrive at a sensible default for 200 hosts as a start point.

Reply:
Mostly done, have a setup similar to your example code now. Can you elaborate somewhat on what you mean by exposing the cherrypy thread pool? Do you think there should be a config option to expand it? Should we grow it ourselves as the number of hosts increases?